File: csharp/SpeedrunKata/stage1/Stage1.cs
namespace SpeedrunKata.stage1
{
    /**
     * Uncomment the tests, and make them pass
     */
    public class Stage1
    {
        public class Items
        {
            public static Items POTATO = new Items(1);
            public static Items TOMATO = new Items(2);
            public static Items LETTUCE = new Items(2);
            public static Items APPLE = new Items(1);
            public static Items BANANA = new Items(3);
            public static Items PEER = new Items(2);

            private readonly int price;
            public Items(int price)
            {
                this.price = price;
            }

            public int GetPrice()
            {
                return this.price;
            }
        }
    }
}

File: rust/src/org/training/speedrun/stage1.rs
use std::ops::Add;

#[derive(Clone,Copy)]
pub enum Items
{
  Potato,
  Tomato,
  Lettuce,
  Apple,
  Banana,
  Peer
}

impl Items
{
  pub fn get_price(&self) -> i32
  {
    match self
    {
      &Items::Potato => 1,
      &Items::Tomato => 2,
      &Items::Lettuce => 2,
      &Items::Apple => 1,
      &Items::Banana => 3,
      &Items::Peer => 2,
    }
  }
}

fn compute_price(v: &std::vec::Vec<Items>) -> i32
{
    v.iter().map(Items::get_price).fold(0i32, Add::add)
}

macro_rules! provide_carts
{
  ( $($name:ident: $value:expr,)*) =>
  {$(
    #[test]
    fn $name() {
      let (v, expected): (std::vec::Vec<Items>, i32) = $value;
      assert_eq!(compute_price(&v), expected);
    }
  )*}
}

provide_carts! {
  cart0: (vec![Items::Apple, Items::Tomato, Items::Banana], 6),
  cart1: (vec![Items::Potato, Items::Potato], 2),
  cart2: (vec![Items::Lettuce, Items::Tomato, Items::Potato], 5),
  cart3: (vec![Items::Apple, Items::Peer, Items::Banana, Items::Peer], 8),
}
